### PR TITLE
KAFKA-5985: update javadoc regarding closing iterators

### PR DESCRIPTION
**Pull request by bbejeck**

### CODE REVIEW

**Comment by bbejeck** [comment]
@miguno @dguy @mjsax @guozhangwang 

Updated Java Docs regarding the need to close `Iterator`s.  Docs to be updated in separate PR.

NM, forgot docs part of this repo as well, will add doc changes to this PR

File: docs/streams/developer-guide.html [hunk 1]
```diff
@@ -185,6 +185,7 @@
         <li>In the <code>init</code> method, schedule the punctuation every 1 second and retrieve the local state store by its name "Counts".</li>
         <li>In the <code>process</code> method, upon each received record, split the value string into words, and update their counts into the state store (we will talk about this feature later in the section).</li>
         <li>In the scheduled <code>punctuate</code> method, iterate the local state store and send the aggregated counts to the downstream processor, and commit the current stream state.</li>
+        <li>When done with the <code>KeyValueIterator&lt;String, Long&gt;</code> we <em>MUST</em> close the iterator, as show above or using the try-with-resources statement (available since JDK7).</li>
```
Comment (mjsax) [inline docs/streams/developer-guide.html]: so we need `<em>` and all upper case for `MUST` ? Seems a little bit too much for me. I would not use all upper-case in general.

Comment (mjsax) [reply docs/streams/developer-guide.html]: So we need to point out JDK7 ? Streams does not support any older version anyway?

File: docs/streams/developer-guide.html [hunk 2]
```diff
@@ -253,6 +254,13 @@
 The default retention setting is <code>Windows#maintainMs()</code> + 1 day. This setting can be overriden by specifying <code>StreamsConfig.WINDOW_STORE_CHANGE_LOG_ADDITIONAL_RETENTION_MS_CONFIG</code> in the <code>StreamsConfig</code>.
 </p>
 
+<p>
+One additional note regarding the use of state stores.  Any time you open an <code>Iterator</code> from a state store you <em>MUST</em> call <code>close</code> on the iterator
```
Comment (mjsax) [inline docs/streams/developer-guide.html]: nit: `close()` (add `()`)

File: docs/streams/developer-guide.html (hunk 1 above)
Comment (mjsax) [reply docs/streams/developer-guide.html]: Also, `we` -> `you` ?

File: docs/streams/developer-guide.html (hunk 2 above)
Comment (mjsax) [reply docs/streams/developer-guide.html]: and `MUST` as above

File: docs/streams/developer-guide.html [hunk 3]
```diff
@@ -253,6 +254,13 @@
...
+when you are done working with it to reclaim resources.  Or you can use the iterator from within a try-with-resources statement (available since JDK7).
```
Comment (mjsax) [inline docs/streams/developer-guide.html]: JDK ?

File: docs/streams/developer-guide.html [hunk 4]
```diff
@@ -253,6 +254,13 @@
...
+    By Not closing an iterator, you may likely encounter an OOM error.
```
Comment (mjsax) [inline docs/streams/developer-guide.html]: `[n]ot`

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueIterator.java [hunk 5]
```diff
@@ -24,7 +24,7 @@
 /**
  * Iterator interface of {@link KeyValue}.
  *
- * Users need to call its {@code close} method explicitly upon completeness to release resources,
+ * Users MUST call its {@code close} method explicitly upon completeness to release resources,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/KeyValueIterator.java]: lower case (same below) not putting a comment everywhere

**Comment by bbejeck** [comment]
updated this.

**Review by mjsax — APPROVED** [review]
LGTM

File: docs/streams/developer-guide.html [hunk 6]
```diff
@@ -185,6 +185,7 @@
         <li>In the <code>init</code> method, schedule the punctuation every 1 second and retrieve the local state store by its name "Counts".</li>
         <li>In the <code>process</code> method, upon each received record, split the value string into words, and update their counts into the state store (we will talk about this feature later in the section).</li>
         <li>In the scheduled <code>punctuate</code> method, iterate the local state store and send the aggregated counts to the downstream processor, and commit the current stream state.</li>
+        <li>When done with the <code>KeyValueIterator&lt;String, Long&gt;</code> you <em>must</em> close the iterator, as show above or using the try-with-resources statement.</li>
```
Comment (miguno) [inline docs/streams/developer-guide.html]: Typo/grammar:

> When done with the <code>KeyValueIterator&lt;String, Long&gt;</code> you <em>must</em> close the iterator, as show above or using the try-with-resources statement.

should be

> When done with the <code>KeyValueIterator&lt;String, Long&gt;</code> you <em>must</em> close the iterator as shown above, or use the try-with-resources statement.

**Review by miguno — CHANGES_REQUESTED** [review]

File: docs/streams/developer-guide.html [hunk 7]
```diff
@@ -185,7 +185,7 @@
         <li>In the <code>init</code> method, schedule the punctuation every 1 second and retrieve the local state store by its name "Counts".</li>
         <li>In the <code>process</code> method, upon each received record, split the value string into words, and update their counts into the state store (we will talk about this feature later in the section).</li>
         <li>In the scheduled <code>punctuate</code> method, iterate the local state store and send the aggregated counts to the downstream processor, and commit the current stream state.</li>
-        <li>When done with the <code>KeyValueIterator&lt;String, Long&gt;</code> you <em>must</em> close the iterator, as show above or using the try-with-resources statement.</li>
+        <li>When done with the <code>KeyValueIterator&lt;String, Long&gt;</code> you <em>must</em> close the iterator, as show above or use the try-with-resources statement.</li>
```
Comment (miguno) [inline docs/streams/developer-guide.html]: Typo: "the iterator, as show above or" -> "the iterator as show[n] above, or "

**Review by miguno — CHANGES_REQUESTED** [review]

**Comment by bbejeck** [comment]
updated

**Review by miguno — APPROVED** [review]

**Comment by bbejeck** [comment]
ping @dguy @guozhangwang for final review and merging

**Review by dguy — APPROVED** [review]
Thanks @bbejeck, LGTM

**Pull request closed without merging**